arangod/Replication/DatabaseInitialSyncer.cpp

```diff
@@ -1009,6 +1244,388 @@ Result DatabaseInitialSyncer::fetchCollectionSync(arangodb::LogicalCollection* c
+      }
+
+      for (std::size_t i = 0; i < rangesSlice.length(); ++i) {
+        VPackSlice masterSlice = rangesSlice.at(i);
+        if (!masterSlice.isArray()) {
+          return Result(
+              TRI_ERROR_REPLICATION_INVALID_RESPONSE,
+              std::string("got invalid response from master at ") +
+                  _config.master.endpoint + batchUrl +
+                  ": response field 'ranges' entry is not a revision range");
+        }
+        auto& currentRange = ranges[chunk];
+        local.seek(std::max(iterResume, static_cast<TRI_voc_rid_t>(currentRange.first)));
+
+        std::size_t removalBound = masterSlice.isEmptyArray()
+                                       ? currentRange.second + 1
+                                       : masterSlice.at(0).getNumber<std::size_t>();
+        TRI_ASSERT(currentRange.first <= removalBound);
+        TRI_ASSERT(removalBound <= currentRange.second + 1);
+        std::size_t mixedBound =
+            masterSlice.isEmptyArray()
+                ? currentRange.second
+                : masterSlice.at(masterSlice.length() - 1).getNumber<std::size_t>();
+        TRI_ASSERT(currentRange.first <= mixedBound);
+        TRI_ASSERT(mixedBound <= currentRange.second);
+
+        while (local.hasMore() && local.revision() < removalBound) {
+          std::string key = transaction::helpers::extractKeyString(local.document());
+          toRemove.emplace_back(key);
+          iterResume = std::max(iterResume, local.revision() + 1);
+          local.next();
+        }
+
+        std::size_t index = 0;
+        while (local.hasMore() && local.revision() <= mixedBound) {
+          TRI_voc_rid_t masterRev = masterSlice.at(index).getNumber<TRI_voc_rid_t>();
+
+          if (local.revision() < masterRev) {
+            std::string key = transaction::helpers::extractKeyString(local.document());
+            toRemove.emplace_back(key);
+            iterResume = std::max(iterResume, local.revision() + 1);
+            local.next();
+          } else if (masterRev < local.revision()) {
+            toFetch.emplace_back(masterRev);
+            ++index;
+            iterResume = std::max(iterResume, masterRev + 1);
+          } else {
+            TRI_ASSERT(local.revision() == masterRev);
+            // match, no need to remove local or fetch from master
+            ++index;
+            iterResume = std::max(iterResume, masterRev + 1);
+            local.next();
+          }
+        }
+        for (; index < masterSlice.length(); ++index) {
+          TRI_voc_rid_t masterRev = masterSlice.at(index).getNumber<TRI_voc_rid_t>();
+          // fetch any leftovers
+          toFetch.emplace_back(masterRev);
+          iterResume = std::max(iterResume, masterRev + 1);
+        }
+
+        while (local.hasMore() &&
+               local.revision() <= std::min(requestResume, currentRange.second)) {
+          std::string key = transaction::helpers::extractKeyString(local.document());
```

Comment:
this seems pretty inefficient, removals can happen by revId only and do not need to do the primary key lookup

Reply:
I know the current physicalCollection API does not have this but we should add it